uat/custom-components/client-python-paho/src/grpc_client_server/grpc_control_server.py

```diff
@@ -0,0 +1,363 @@
+from exceptions.mqtt_exception import MQTTException
+
+PORT_MIN = 1
+PORT_MAX = 65_535
+
+KEEPALIVE_OFF = 0
+KEEPALIVE_MIN = 5
+KEEPALIVE_MAX = 65_535
+
+TIMEOUT_MIN = 1
+
+
+class GRPCControlServer(mqtt_client_control_pb2_grpc.MqttClientControlServicer):
+    """Implementation of gRPC server handles requests of OTF."""
+
+    logger = logging.getLogger("GRPCControlServer")
+
+    def __init__(self, client: GRPCDiscoveryClient, address: str):
+        """
+        Construct GRPCControlServer
+        Parameters
+        ----------
+        client - GRPCDiscoveryClient object
+        address - local gRPC service address
+        """
+        self.__logger = GRPCControlServer.logger
+        self.__shutdown_reason = ""
+        self.__stop_server_future = asyncio.Future()
+        self.__mqtt_lib: MQTTLib = None
+
+        # TODO delete suppress warning
+        self.__client = client  # pylint: disable=unused-private-member
+
+        super().__init__()
+
+        self.__server = grpc.aio.server()
+        mqtt_client_control_pb2_grpc.add_MqttClientControlServicer_to_server(self, self.__server)
+        self.__chosen_port = self.__server.add_insecure_port(address)
+
+    def get_port(self) -> int:
+        """Returns gRPC server port"""
+        return self.__chosen_port
+
+    def get_shutdown_reason(self) -> str:
+        """Returns shutdown reason"""
+        return self.__shutdown_reason
+
+    async def start_grpc_server(self):
+        """Start grpc server"""
+        await self.__server.start()
+
+    async def wait(self, mqtt_lib: MQTTLib):
+        """
+        Wait until incoming shutdown request
+        Parameters
+        ----------
+        mqtt_lib - MQTT side of the client to handler incoming requests
+        """
+        self.__mqtt_lib = mqtt_lib
+        self.__logger.info("Server awaiting termination")
+        await self.__stop_server_future
+        await self.__server.stop(None)
+        await self.__server.wait_for_termination()
+        self.__logger.info("Server termination done")
+
+    def unblock_wait(self):
+        """Unblock wait method"""
+        try:
+            self.__stop_server_future.set_result(True)
+        except asyncio.InvalidStateError:
+            pass
+
+    def close(self):
+        """Closes the gRPC server."""
+        self.unblock_wait()
+
+    async def ShutdownAgent(self, request: ShutdownRequest, context: grpc.aio.ServicerContext) -> Empty:
+        """
+        override
+        Handler of ShutdownAgent gRPC call.
+        Parameters
+        ----------
+        request - incoming request
+        context - request context
+        Returns Empty object
+        """
+        self.__shutdown_reason = request.reason
+        self.__logger.info("shutdownAgent: reason %s", self.__shutdown_reason)
+        self.unblock_wait()
+        return Empty()
+
+    async def CreateMqttConnection(  # pylint: disable=too-many-locals,too-many-branches,too-many-statements

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 54 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 98% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code block on lines 185-241 into a separate function.